File: rtsp_connection.cpp
#include <string.h>
#include <time.h>

#include <string>
#include <algorithm>

#include "logging.h"
#include "rtsp_connection.h"
#include "tcp_server.h"
#include "session.h"
#include "media.h"

/*----------------------------------------------------------------------------*/
static std::string trim(const std::string str)
{
    const std::size_t begin = str.find_first_not_of(" \t");
    const std::size_t end = str.find_last_not_of(" \t");
    if(begin == std::string::npos) {
	return "";
    }
    return str.substr(begin, end - begin +1);
}


/*----------------------------------------------------------------------------*/
static char ascii2lower(char in) 
{
    if ((in <= 'Z') && (in >= 'A')) {
	return in - 'Z' + 'z';
    }
    return in;
}


/*----------------------------------------------------------------------------*/
static bool split_name_value_pair(const std::string & str, std::string & name,
	       std::string & value)
{
    bool retVal = false;

    const std::size_t colon = str.find(":");
    if(colon != std::string::npos) {
	name = std::string(str, 0, colon);
	std::transform(name.begin(), name.end(), name.begin(), ascii2lower);
	value = trim(str.substr(colon+1));
	LOG_DEBUG("%s => %s", name.c_str(), value.c_str());
	retVal = true;
    }
    return retVal;
}


/*----------------------------------------------------------------------------*/
const std::string extract_path(const std::string url)
{
    if(url.rfind("rtsp://", 0) == 0) {
	const std::size_t idx = url.find("/",7);
	return url.substr(idx+1);
    }
    return "";
}



/*----------------------------------------------------------------------------*/
RtspConnection::RtspConnection() :  mParsingState(PARSING_REQUEST_LINE)
{
}


/*----------------------------------------------------------------------------*/
in_port_t RtspConnection::get_rtsp_server_port() const
{
    return dynamic_cast<TcpServer *>(get_network())->get_listening_port();
}


/*----------------------------------------------------------------------------*/
int RtspConnection::parse_recv(const Byte * pData, unsigned len)
{
    int retVal = 0;

    const char * p = reinterpret_cast<const char *>(pData);
    const char * end = reinterpret_cast<const char *>(memmem(reinterpret_cast<const void *>(p), len, "\r\n\r\n", 4));

    if(end) {
	const unsigned size = (end - p);
    	unsigned left = size;
        
	// Split up into lines to parse
	while(left > 0) {
	    while(((*p == '\r') || (*p == '\n')) && (left > 0)) {
	    	++p;
	    	left--;
	    }

	    const char * q = p;
	    while((*q != '\r') && (*q != '\n') && (left > 0)) {
	    	++q;
	    	left--;
	    }
	    const unsigned lineLen = q - p;
	    if(lineLen > 0) {
	        std::string str(p, lineLen);
	        parse_line(str);
	    }
	    p = q;
        }	  
//	LOG_INFO("%s\n", pData);
	generate_response();
	retVal = size + 4;
    } else {
        LOG_DEBUG("%i %i %i %i", pData[len-4], pData[len-3], pData[len-2], pData[len-1]);
    }
    return retVal;
}


/*----------------------------------------------------------------------------*/
void RtspConnection::generate_response()
{
     std::string response = get_response();
     send(reinterpret_cast<const unsigned char *>(response.c_str()),
		    response.size()); 
}



/*----------------------------------------------------------------------------*/
void RtspConnection::parse_line(const std::string & str)
{
    LOG_DEBUG("Line = %s", str.c_str());	

    switch(mParsingState) {
	case PARSING_REQUEST_LINE:
	    parse_request_line(str);
	    break;
	case PARSING_OPTIONS_REQUEST:
	    parse_options_request(str);
	    break;
	case PARSING_DESCRIBE_REQUEST:
	    parse_describe_request(str);
	    break;
	case PARSING_SETUP_REQUEST:
	    parse_setup_request(str);
	    break;
	case PARSING_PLAY_REQUEST:
	    parse_play_request(str);
	    break;
    }
}


/*----------------------------------------------------------------------------*/
std::string RtspConnection::get_response()
{
    std::string retVal;

    switch(mParsingState) {
	case PARSING_REQUEST_LINE:
	    break;
	case PARSING_OPTIONS_REQUEST:
	    retVal = generate_options_response();
	    break;
	case PARSING_DESCRIBE_REQUEST:
	    retVal = generate_describe_response();
	    break;
	case PARSING_SETUP_REQUEST:
	    retVal = generate_setup_response();
	    break;
	case PARSING_PLAY_REQUEST:
	    retVal = generate_play_response();
	    break;
    }
    mParsingState = PARSING_REQUEST_LINE;
    LOG_DEBUG("%s", retVal.c_str());
    return retVal;
}



/*----------------------------------------------------------------------------*/
void RtspConnection::parse_request_line(const std::string & str)
{
    std::string str1 = trim(str);
    std::size_t idx1 = str1.find_first_of(" \t");
    std::size_t idx2 = str1.find_last_of(" \t");

    if((idx1 != std::string::npos) && (idx2 != std::string::npos)) {
	std::string request = str1.substr(0, idx1);
	mUrl = trim(str1.substr(idx1, idx2-idx1));
	std::string version = str1.substr(idx2+1);

//	LOG_DEBUG("Request = %s", request.c_str());
//	LOG_DEBUG("URL = %s", mUrl.c_str());
//	LOG_DEBUG("Version = %s", version.c_str());

	if(request == "OPTIONS") {
	    mParsingState = PARSING_OPTIONS_REQUEST;
	} else if(request == "DESCRIBE") {
	    mParsingState = PARSING_DESCRIBE_REQUEST;
	} else if(request == "SETUP") {
	    mParsingState = PARSING_SETUP_REQUEST;
	} else if(request == "PLAY") {
	    mParsingState = PARSING_PLAY_REQUEST;
	} else {
	    LOG_ERROR("Unknown request %s", request.c_str());
	    // TODO
	}
    }
}



/*----------------------------------------------------------------------------*/
void RtspConnection::parse_options_request(const std::string & str)
{
    std::string name;
    std::string value;

    if(split_name_value_pair(str, name, value)) {
	// CSeq
	if(name == "cseq") {
	    mCseq = value;
        }
    }	
}


/*----------------------------------------------------------------------------*/
std::string RtspConnection::generate_options_response()
{
    return std::string("RTSP/1.0 200 OK\r\n" \
	"CSeq: ") + mCseq + "\r\n" \
	"Public: DESCRIBE, SETUP, PLAY, TEARDOWN\r\n\r\n";
}


/*----------------------------------------------------------------------------*/
void RtspConnection::parse_describe_request(const std::string & str)
{
    std::string name;
    std::string value;

    if(split_name_value_pair(str, name, value)) {
	// CSeq
	if(name == "cseq") {
	    mCseq = value;
        } else if(name == "accept") {
	    // Check that application/sdp is an option
	    if(value.find("application/sdp") == std::string::npos) {
		// TODO
            }
	}
    }	
}


/*----------------------------------------------------------------------------*/
std::string RtspConnection::generate_describe_response()
{
    std::string pathname = extract_path(mUrl);
    Session * pSession = Media::get_session(pathname.c_str());
    // Mandatory fields:
    // CSeq:
    // Content-type:
    // Content-Length;

    // SDP fields
    // v= (protocol version)
    // o = (origintator and session identifier)
    // s = (session name)
    // t = (time the session is active)
    // m = (media name and transport address)
    std::string sdp = "v=0\r\n" \
	    "o=- " + std::to_string(pSession->get_sdp_id()) + " " 
	    	+ std::to_string(pSession->get_sdp_ver()) + " IN IP4 "
	       	+ get_hostip() + "\r\n" \
	    "s=TV Session\r\n" \
	    "i=Sound from the TV\r\n" \
	    "t= 0 0\r\n" \
	    "a=recvonly\r\n" \
	    "a=control:rtsp://" + get_hostname() + ":" 
	    	+ std::to_string(get_rtsp_server_port()) + "/"
	       	+ pSession->get_pathname() + "\r\n" \
	    "m=audio " + std::to_string(pSession->get_our_rtp_port()) 
	    	+ " RTP/AVP 31\r\n" \
	    "a=rtpmap:31 OPUS/48000/2\r\n";

    return std::string("RTSP/1.0 200 OK\r\n"\
	    "CSeq: ") + mCseq + "\r\n" \
	    "Content-Type: application/sdp\r\n" \
	    "Content-Length: " + std::to_string(sdp.size()) + "\r\n\r\n" 
	    + sdp;
}


/*----------------------------------------------------------------------------*/
void RtspConnection::parse_setup_request(const std::string & str)
{
    std::string pathname = extract_path(mUrl);
    Session * pSession = Media::get_session(pathname.c_str());

    std::string name;
    std::string value;

    if(split_name_value_pair(str, name, value)) {
	// CSeq
	if(name == "cseq") {
	    mCseq = value;
        } else if(name == "transport") {
  	    std::size_t idx = value.find(";");
	    std::string proto;

	    if(idx != std::string::npos) {
		proto = value.substr(0, idx);
	    } else {
		proto = value;
	    }
	    LOG_DEBUG("proto=%s", proto.c_str());
	
	    // Transport fields are separated with semi-colons.
	    while(idx != std::string::npos) {
		const std::size_t idx2 = value.find(";", idx+1);
		const std::size_t idx3 = value.find("=", idx+1);

		std::string subName;
		std::string subValue;
		if(idx2 != std::string::npos) {
		    if((idx3 != std::string::npos) && (idx3 < idx2)) {
 		        subName = value.substr(idx+1, idx3-idx-1);
			subValue = value.substr(idx3+1, idx2-idx3-1);
		    } else {
			subName = value.substr(idx+1, idx2-idx-1);
		    }
		} else if(idx3 != std::string::npos) {
 		    subName = value.substr(idx+1, idx3-idx-1);
		    subValue = value.substr(idx3+1);
		} else {
		    subName = value.substr(idx+1);
		}
		idx = idx2;
		LOG_DEBUG("%s=%s", subName.c_str(), subValue.c_str());
		if(subName == "client_port") {
		    const std::size_t idx4 = subValue.find("-");
		    if(idx4 != std::string::npos) {
			pSession->set_peer_rtp_port(std::stoi(subValue.substr(0, idx4)));
			pSession->set_peer_rtcp_port(std::stoi(subValue.substr(idx4+1)));
		    } else {
			in_port_t port = std::stoi(subValue);
			pSession->set_peer_rtp_port(port);
			pSession->set_peer_rtcp_port(port+1);
		    }
		}
            }
	}
    }	
}


/*----------------------------------------------------------------------------*/
std::string RtspConnection::generate_setup_response()
{
    std::string pathname = extract_path(mUrl);
    Session * pSession = Media::get_session(pathname.c_str());
    // Mandatory fields:
    // CSeq:
    // Session:
    // Transport:
    std::string transport = "RTP/AVP/UDP;unicast;" \
	"client_port=" + std::to_string(pSession->get_peer_rtp_port()) + "-"
       	+ std::to_string(pSession->get_peer_rtcp_port()) + ";"\
	"server_port=" + std::to_string(pSession->get_our_rtp_port()) + "-"
	+ std::to_string(pSession->get_our_rtcp_port());

    return std::string("RTSP/1.0 200 OK\r\n" \
        "CSeq: ") + mCseq + "\r\n" \
	"Session: " + pSession->get_id() + "\r\n" \
	"Transport: " + transport + "\r\n\r\n";
}


/*----------------------------------------------------------------------------*/
void RtspConnection::parse_play_request(const std::string & str)
{
//    std::string pathname = extract_path(mUrl);
//    Session * pSession = Media::get_session(pathname.c_str());

    std::string name;
    std::string value;

    if(split_name_value_pair(str, name, value)) {
	// CSeq
	if(name == "cseq") {
	    mCseq = value;
        } else if(name == "range") {
	}
    }	
}


/*----------------------------------------------------------------------------*/
std::string RtspConnection::generate_play_response()
{
    // Mandatory fields:
    // CSeq:
    
    return std::string("RTSP/1.0 200 OK\r\n" \
        "CSeq: ") + mCseq + "\r\n\r\n";
}

File: py_src/rtsp_connection.py
import socket

try:
	from . import media
	from . import network_interfaces
	from .session import ChangeNotAllowedError
except ImportError:
	import media
	import network_interfaces
	from session import ChangeNotAllowedError

class RtspConnection(object):

	def __init__(self, connfd, peer_address, listening_port):
		self.connfd = connfd
		self.read_buf = b''
		self.peer_address = peer_address
		self.listening_port = listening_port

	def fileno(self):
		return self.connfd.fileno()


	def read(self):
		new_data = self.connfd.recv(512)
		if not new_data:
			return False
		read_buf = self.read_buf + new_data
		if read_buf[-4:] != b'\r\n\r\n':
			self.read_buf = read_buf
			return True
		self.read_buf = b''
		lines = read_buf[:-4].splitlines()
		return self.parse_request(lines)

	def parse_request(self, lines):
		tokens = lines[0].split()
		request = tokens[0]
		version = tokens[-1]
		url = b" ".join(tokens[1:-1])
		if url.startswith(b"rtsp://"):
			url = url[7:]

		attributes = {}
		for line in lines[1:]:
			name, value = line.split(b":", 1)
			name = name.strip()
			value = value.strip()
			attributes[name.lower()] = value
		print("Recv: ", request)
		print(attributes)
		if request == b"OPTIONS":
			return self.parse_option_request(attributes)
		if request == b"DESCRIBE":
			return self.parse_describe_request(url, attributes)
		if request == b"SETUP":
			return self.parse_setup_request(url, attributes)
		if request == b"PLAY":
			return self.parse_play_request(url, attributes)
		if request == b"TEARDOWN":
			return self.parse_teardown_request(url, attributes)
		if request in (b"RECORD", b"ANNOUNCE",
				b"GET_PARAMETER", b"PAUSE", b"REDIRECT",
				b"SET_PARAMETER"):
			return self.Error(405, attributes)
		return self.Error(400, attributes)

	def keep_alive(self, attributes):
		if b"connection" in attributes:
			value = attributes[b"connection"]
			if value.lower() == b"close":
				return False
		return True

	def parse_option_request(self, attributes):
		if b"require" in attributes:
			return self.Error(551, attributes)
		if b"proxy-require" in attributes:
			return self.Error(551, attributes)
		lines = [
			b"RTSP/1.0 200 OK",
			b"CSeq: " + attributes[b"cseq"],
			b"Public: DESCRIBE, SETUP, PLAY, TEARDOWN",
			b"", b""]

		print(lines)
		self.connfd.send(b"\r\n".join(lines))
		return self.keep_alive(attributes)


	def parse_describe_request(self, url, attributes):
		if b"require" in attributes:
			return self.Error(551, attributes)
		if b"proxy-require" in attributes:
			return self.Error(551, attributes)

		session = media.get_session(url)

#    // Mandatory fields:
#    // CSeq:
#    // Content-type:
#    // Content-Length;

#    // SDP fields
#    // v= (protocol version)
#    // o = (origintator and session identifier)
#    // s = (session name)
#    // t = (time the session is active)
#    // m = (media name and transport address)

		pt = str(session.get_payload_type())
		sdp = [
			"v=0",
			"o=- {0} {1} IN IP4 {2}".format(session.get_sdp_id(),
			session.get_sdp_ver(), self.get_hostip(4)),
			"s=TV Session",
			"i=Sound from the TV",
			"t= 0 0",
			"a=recvonly",
			"a=control:rtsp://{0}:{1}/{2}".format(self.get_hostname(),
			self.listening_port, session.get_pathname()),
			"m=audio {0} RTP/AVP {1}".format(session.get_our_rtp_port(), pt),
			"a=rtpmap:{0} opus/{1}/{2}".format(pt, session.get_raw_bit_rate(),
			session.get_num_of_channels()),
			""]
		sdp = "\r\n".join(sdp)
		sdp = sdp.encode(encoding = "UTF-8")

# Content-Encoding not needed as entity is in plain text
# Content-Language not needed as entity is applicabel to all
		lines = [
			b"RTSP/1.0 200 OK",
			b"CSeq: " + attributes[b"cseq"],
			b"Content-Type: application/sdp",
			b"Content-Length: " + str(len(sdp)).encode(encoding="UTF-8"),
			b"", b""]
		self.connfd.send(b"\r\n".join(lines))
		self.connfd.send(sdp)
		return self.keep_alive(attributes)


	def Error(self, num, attributes):
		msgs = {
			400 : "Bad Request",
			405 : "Method Not Allowed",
			454 : "Session Not Found",
			455 : "Method Not Valid In This State",
			459 : "Aggregate Operation Not Allowed",
			461 : "Unsupported Transport",
			551 : "Option not supported",
		}
		lines = [
			"RTSP/1.0 {0} {1}".format(num, msgs[num]).encode(encoding = "UTF-8"),
			b"CSeq: " + attributes[b"cseq"],
		]
		if num == 405:
			lines.append(b"Allow: DESCRIBE, SETUP, PLAY, TEARDOWN")
		if num == 551:
			if b"require" in attributes:
				unsupported = attributes[b"require"]
			else:
				unsupported = attributes[b"proxy-require"]
			lines.append(b"Unsupported: {0}".format(unsupported).encode(encoding = "UTF-8"))
		lines += [
			b"", 
			b""]
		print(lines)
		self.connfd.send(b"\r\n".join(lines))
		return self.keep_alive(attributes)


	def parse_setup_request(self, url, attributes):
		if b"require" in attributes:
			return self.Error(551, attributes)
		if b"proxy-require" in attributes:
			return self.Error(551, attributes)
		session = media.get_session(url)

		value = attributes[b"transport"]
		# We dont do aggregated sessions...
		if b"session" in attributes:
			return self.Error(459, attributes)

		tokens = value.split(b";")
		for token in tokens:
			token = token.strip()

		proto = tokens[0].strip()

		if proto.upper() not in (b"RTP/AVP", b"RTP/AVP/UDP"):
			return self.Error(461, attributes)

		print("proto=", proto)
#        print(tokens)

		values = {}
		for token in tokens[1:]:
			idx = token.find(b"=")
			if idx > 0:
				name, value = token[:idx].strip().lower(), token[idx+1:].strip()
			else:
				name = token.strip().lower()
				value = True
			values[name] = value

		ports = values[b"client_port"].split(b"-")
		try:
			if len(ports) > 1:
				session.set_peer_rtp_port(int(ports[0]))
				session.set_peer_rtcp_port(int(ports[1]))
			else:
				session.set_peer_rtp_port(int(ports[0]))
				session.set_peer_rtcp_port(int(ports[0])+1)

			if self.peer_address.find(".") >= 0:
				session.set_our_address(self.get_hostip(4))
			else:
				session.set_our_address(self.get_hostip(6))

			session.add_peer_address(self.peer_address);
		except ChangeNotAllowedError as err:
			return self.Error(455, attributes)

#    // Mandatory fields:
#    // CSeq:
#    // Session:
#    // Transport:
		transport = "RTP/AVP/UDP;unicast;client_port=" \
            + str(session.get_peer_rtp_port()) + "-" + str(session.get_peer_rtcp_port()) \
            + ";server_port=" + str(session.get_our_rtp_port()) + "-" \
            + str(session.get_our_rtcp_port())
		transport = transport.encode(encoding="UTF-8")

		lines = [
			b"RTSP/1.0 200 OK",
			b"CSeq: " + attributes[b"cseq"],
			b"Session: " + session.get_rtp_id(),
			b"Transport: " + transport,
			b"", b""]

		print(lines)
		self.connfd.send(b"\r\n".join(lines))
		return self.keep_alive(attributes)


	def parse_play_request(self, url, attributes):
		if b"require" in attributes:
			return self.Error(551, attributes)
		if b"proxy-require" in attributes:
			return self.Error(551, attributes)
		session_id = None
		if b"session" in attributes:
			session_id = attributes[b"session"]
		print(session_id)
		session = media.get_session(url)
		if session.get_rtp_id() != session_id:
			return self.Error(454, attributes)

		session.play()

#    // Mandatory fields:
#    // CSeq:
		lines = [
			b"RTSP/1.0 200 OK",
			b"CSeq: " + attributes[b"cseq"],
			b"Session: " + session.get_rtp_id(),
			b"", b""]

		print(lines)
		self.connfd.send(b"\r\n".join(lines))
		return self.keep_alive(attributes)


	def parse_teardown_request(self, url, attributes):
		if b"require" in attributes:
			return self.Error(551, attributes)
		if b"proxy-require" in attributes:
			return self.Error(551, attributes)
		session_id = None
		if b"session" in attributes:
			session_id = attributes[b"session"]
		session = media.get_session(url)
		if session.get_rtp_id() != session_id:
			return self.Error(454, attributes)
		session.stop()
		return self.keep_alive(attributes)


	def get_hostip(self, ver):
		return network_interfaces.get_hostip(self.connfd, ver)

	def get_hostname(self):
		return socket.getfqdn()
